### PR TITLE
8263636: Add --disable-registry option to jhsdb debugd

### PR DESCRIPTION
**Pull request by YaSuenag**
`jhsdb debugd` will start RMI registry by default, but we want to prevent it due to port confliction in some cases. We can control it with `sun.jvm.hotspot.rmi.startRegistry` system property. However we have no way to set it excepting system property. jhsdb should provide the way to set it as a command line option.

This PR introduces `--disableregistry` to `jhsdb debugd`. Please review [CSR](https://bugs.openjdk.java.net/browse/JDK-8264021) too.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8263636](https://bugs.openjdk.java.net/browse/JDK-8263636): Add --disable-registry option to jhsdb debugd

### Reviewers
 * [Chris Plummer](https://openjdk.java.net/census#cjplummer) (@plummercj - **Reviewer**)
 * [Kevin Walls](https://openjdk.java.net/census#kevinw) (@kevinjwalls - Committer)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/3233/head:pull/3233` \
`$ git checkout pull/3233`

Update a local copy of the PR: \
`$ git checkout pull/3233` \
`$ git pull https://git.openjdk.java.net/jdk pull/3233/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 3233`

View PR using the GUI difftool: \
`$ git pr show -t 3233`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/3233.diff">https://git.openjdk.java.net/jdk/pull/3233.diff</a>

</details>

### CODE REVIEW

File: src/jdk.hotspot.agent/share/classes/sun/jvm/hotspot/SALauncher.java
Comment (plummercj) [inline src/jdk.hotspot.agent/share/classes/sun/jvm/hotspot/SALauncher.java:89]: You change here makes it look like if you specify `id@` then you also need to specify the port. I'd suggest also including the original line that just has `id@debugserver`.

File: src/jdk.hotspot.agent/share/classes/sun/jvm/hotspot/SALauncher.java
Comment (plummercj) [inline src/jdk.hotspot.agent/share/classes/sun/jvm/hotspot/SALauncher.java:103]: "to use RMI registry that exists" doesn't read well. Do you mean "use already started RMI registry"?

File: src/jdk.hotspot.agent/share/man/jhsdb.1
Comment (plummercj) [inline src/jdk.hotspot.agent/share/man/jhsdb.1:188]: Is this what you mean: `Otherwise it will not be started, and the already started RMI registry will be used instead.`

File: test/hotspot/jtreg/serviceability/sa/sadebugd/DebugdRmidTest.java
Comment (plummercj) [inline test/hotspot/jtreg/serviceability/sa/sadebugd/DebugdRmidTest.java:58]: "Sleep 3 sec to wait for rmid to start".

File: test/hotspot/jtreg/serviceability/sa/sadebugd/DebugdRmidTest.java
Comment (plummercj) [inline test/hotspot/jtreg/serviceability/sa/sadebugd/DebugdRmidTest.java:37]: Can you add an `@summary` comment?

File: test/hotspot/jtreg/serviceability/sa/sadebugd/DebugdUtils.java
Comment (plummercj) [inline test/hotspot/jtreg/serviceability/sa/sadebugd/DebugdUtils.java:41]: I'd suggest removing the existing empty lines rather than follow the pattern of adding them.

**Review by plummercj — CHANGES_REQUESTED** [review]
Overall the changes look good. I just have some minor suggestions for comments, code formatting, and help output.

File: src/jdk.hotspot.agent/share/classes/sun/jvm/hotspot/SALauncher.java
Comment (YaSuenag) [reply src/jdk.hotspot.agent/share/classes/sun/jvm/hotspot/SALauncher.java:90]: I reverted the original line in new commit.

We can also specify port number without `id@` (e.g. `--connect debugserver:1234`). Is it ok not to describe on help message? IMHO it is not good to describe all patterns because it might be verbosely.

File: test/hotspot/jtreg/serviceability/sa/sadebugd/DebugdRmidTest.java
Comment (YaSuenag) [reply test/hotspot/jtreg/serviceability/sa/sadebugd/DebugdRmidTest.java:51]: They are needed.

If we don't give `execPolicy=none`, we can see warning on console.
SA code would run on rmid, so we need to add module.

File: src/jdk.hotspot.agent/share/classes/sun/jvm/hotspot/SALauncher.java
Comment (YaSuenag) [reply src/jdk.hotspot.agent/share/classes/sun/jvm/hotspot/SALauncher.java:88]: > On the other hand, maybe you could go with just the id@debugserver:1234 example and leave the other 3 off. The syntax does clearly show the id and port are optional.

Agree, I left `id@debugserver:1234` only in new commit.

File: test/hotspot/jtreg/serviceability/sa/sadebugd/DebugdRmidTest.java
Comment (plummercj) [reply test/hotspot/jtreg/serviceability/sa/sadebugd/DebugdRmidTest.java:51]: I was actually just referring to `--add-modules`, but github added a few lines before the one I selected.

I guess I'm not fully understanding what `rmid` is for in this context, and how it relates to the `rmiregistry` command. I thought it was starting the registry, but that does not seem to be it's primary purpose (maybe it starts it as a side affect). Also, it is deprecated.

Comment (YaSuenag) [reply test/hotspot/jtreg/serviceability/sa/sadebugd/DebugdRmidTest.java:51]: SA has RMI remote object, it will be handled in RMI registry.
We can see following exception without `--add-modules`.

```
$ jhsdb debugd --pid 40339 --disableregistry --hostname localhost --registryport 1098
Attaching to process ID 40339 and starting RMI services, please wait...
Error attaching to process or starting server: sun.jvm.hotspot.debugger.DebuggerException: java.rmi.ServerException: RemoteException occurred in server thread; nested exception is:
        java.rmi.UnmarshalException: error unmarshalling arguments; nested exception is:
        java.lang.ClassNotFoundException: sun.jvm.hotspot.debugger.remote.RemoteDebugger
        at jdk.hotspot.agent/sun.jvm.hotspot.RMIHelper.rebind(RMIHelper.java:75)
        at jdk.hotspot.agent/sun.jvm.hotspot.HotSpotAgent.setupDebugger(HotSpotAgent.java:380)
        at jdk.hotspot.agent/sun.jvm.hotspot.HotSpotAgent.go(HotSpotAgent.java:330)
        at jdk.hotspot.agent/sun.jvm.hotspot.HotSpotAgent.startServer(HotSpotAgent.java:216)
        at jdk.hotspot.agent/sun.jvm.hotspot.SALauncher.runDEBUGD(SALauncher.java:437)
        at jdk.hotspot.agent/sun.jvm.hotspot.SALauncher.main(SALauncher.java:499)
Caused by: java.rmi.ServerException: RemoteException occurred in server thread; nested exception is:
        java.rmi.UnmarshalException: error unmarshalling arguments; nested exception is:
        java.lang.ClassNotFoundException: sun.jvm.hotspot.debugger.remote.RemoteDebugger
        at java.rmi/sun.rmi.server.UnicastServerRef.dispatch(UnicastServerRef.java:389)
        at java.rmi/sun.rmi.transport.Transport$1.run(Transport.java:200)
        at java.rmi/sun.rmi.transport.Transport$1.run(Transport.java:197)
        at java.base/java.security.AccessController.doPrivileged(AccessController.java:691)
        at java.rmi/sun.rmi.transport.Transport.serviceCall(Transport.java:196)
        at java.rmi/sun.rmi.transport.tcp.TCPTransport.handleMessages(TCPTransport.java:587)
        at java.rmi/sun.rmi.transport.tcp.TCPTransport$ConnectionHandler.run0(TCPTransport.java:828)
        at java.rmi/sun.rmi.transport.tcp.TCPTransport$ConnectionHandler.lambda$run$0(TCPTransport.java:705)
        at java.base/java.security.AccessController.doPrivileged(AccessController.java:391)
        at java.rmi/sun.rmi.transport.tcp.TCPTransport$ConnectionHandler.run(TCPTransport.java:704)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1135)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
        at java.base/java.lang.Thread.run(Thread.java:831)
        at java.rmi/sun.rmi.transport.StreamRemoteCall.exceptionReceivedFromServer(StreamRemoteCall.java:303)
        at java.rmi/sun.rmi.transport.StreamRemoteCall.executeCall(StreamRemoteCall.java:279)
        at java.rmi/sun.rmi.server.UnicastRef.invoke(UnicastRef.java:380)
        at java.rmi/sun.rmi.registry.RegistryImpl_Stub.rebind(RegistryImpl_Stub.java:158)
        at java.rmi/java.rmi.Naming.rebind(Naming.java:177)
        at jdk.hotspot.agent/sun.jvm.hotspot.RMIHelper.rebind(RMIHelper.java:64)
        ... 5 more
Caused by: java.rmi.UnmarshalException: error unmarshalling arguments; nested exception is:
        java.lang.ClassNotFoundException: sun.jvm.hotspot.debugger.remote.RemoteDebugger
        at java.rmi/sun.rmi.registry.RegistryImpl_Skel.dispatch(RegistryImpl_Skel.java:157)
        at java.rmi/sun.rmi.server.UnicastServerRef.oldDispatch(UnicastServerRef.java:466)
        at java.rmi/sun.rmi.server.UnicastServerRef.dispatch(UnicastServerRef.java:296)
        at java.rmi/sun.rmi.transport.Transport$1.run(Transport.java:200)
        at java.rmi/sun.rmi.transport.Transport$1.run(Transport.java:197)
        at java.base/java.security.AccessController.doPrivileged(AccessController.java:691)
        at java.rmi/sun.rmi.transport.Transport.serviceCall(Transport.java:196)
        at java.rmi/sun.rmi.transport.tcp.TCPTransport.handleMessages(TCPTransport.java:587)
        at java.rmi/sun.rmi.transport.tcp.TCPTransport$ConnectionHandler.run0(TCPTransport.java:828)
        at java.rmi/sun.rmi.transport.tcp.TCPTransport$ConnectionHandler.lambda$run$0(TCPTransport.java:705)
        at java.base/java.security.AccessController.doPrivileged(AccessController.java:391)
        at java.rmi/sun.rmi.transport.tcp.TCPTransport$ConnectionHandler.run(TCPTransport.java:704)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1135)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
        at java.base/java.lang.Thread.run(Thread.java:831)
Caused by: java.lang.ClassNotFoundException: sun.jvm.hotspot.debugger.remote.RemoteDebugger
        at java.base/java.net.URLClassLoader.findClass(URLClassLoader.java:432)
        at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:586)
        at java.rmi/sun.rmi.server.LoaderHandler$Loader.loadClass(LoaderHandler.java:1207)
        at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:519)
        at java.base/java.lang.Class.forName0(Native Method)
        at java.base/java.lang.Class.forName(Class.java:466)
        at java.rmi/sun.rmi.server.LoaderHandler.loadClassForName(LoaderHandler.java:1221)
        at java.rmi/sun.rmi.server.LoaderHandler.loadProxyInterfaces(LoaderHandler.java:731)
        at java.rmi/sun.rmi.server.LoaderHandler.loadProxyClass(LoaderHandler.java:674)
        at java.rmi/sun.rmi.server.LoaderHandler.loadProxyClass(LoaderHandler.java:611)
        at java.rmi/java.rmi.server.RMIClassLoader$2.loadProxyClass(RMIClassLoader.java:646)
        at java.rmi/java.rmi.server.RMIClassLoader.loadProxyClass(RMIClassLoader.java:311)
        at java.rmi/sun.rmi.server.MarshalInputStream.resolveProxyClass(MarshalInputStream.java:254)
        at java.base/java.io.ObjectInputStream.readProxyDesc(ObjectInputStream.java:1944)
        at java.base/java.io.ObjectInputStream.readClassDesc(ObjectInputStream.java:1886)
        at java.base/java.io.ObjectInputStream.readOrdinaryObject(ObjectInputStream.java:2196)
        at java.base/java.io.ObjectInputStream.readObject0(ObjectInputStream.java:1706)
        at java.base/java.io.ObjectInputStream.readObject(ObjectInputStream.java:496)
        at java.base/java.io.ObjectInputStream.readObject(ObjectInputStream.java:454)
        at java.rmi/sun.rmi.registry.RegistryImpl_Skel.dispatch(RegistryImpl_Skel.java:154)
        ... 14 more
```

As an option, we can also use RMI registry which is started by debugd as following:

* console 1
    * `jhsdb debugd --disableregistry --pid 40859`
* console 2
    * `jhsdb -J-Dsun.jvm.hotspot.rmi.serverNamePrefix=second debugd --disableregistry --pid 40859`

Comment (plummercj) [reply test/hotspot/jtreg/serviceability/sa/sadebugd/DebugdRmidTest.java:51]: Both of the above are using `--disableregistry`. Is that what you meant to do? I would think that you would not want that on the first one.

Comment (YaSuenag) [reply test/hotspot/jtreg/serviceability/sa/sadebugd/DebugdRmidTest.java:51]: Sorry, the correct commands are as follows:

* console 1 (start RMI registry)
    * `jhsdb debugd --pid 40859`
* console 2 (add `--disableregistry` to use it in console 1)
    * `jhsdb -J-Dsun.jvm.hotspot.rmi.serverNamePrefix=second debugd --disableregistry --pid 40860`

Comment (plummercj) [reply test/hotspot/jtreg/serviceability/sa/sadebugd/DebugdRmidTest.java:51]: Is this the actual use case you are trying to address (registry already started by a debugd process). If yes, then I think that's the best approach for the test. If no, what is the use case? If it's some random application starting the registry, will there be issue related to it not having been started with `--add-modules=jdk.hotspot.agent`?

Comment (YaSuenag) [reply test/hotspot/jtreg/serviceability/sa/sadebugd/DebugdRmidTest.java]: I updated testcase. Could you review again?

Comment (plummercj) [reply test/hotspot/jtreg/serviceability/sa/sadebugd/DebugdRmidTest.java]: Looks good.

**Review by plummercj — APPROVED** [review]

**Comment by YaSuenag** [comment]
Thanks @plummercj for the review!
I updated PR to rename `--disableregistry` to `--disable-registry` which is suggested in CSR.

I will finalize CSR when it is reviewed.

**Review by plummercj — APPROVED** [review]

**Comment by YaSuenag** [comment]
Ping: Could you review this change? I need one more reviewer to push. CSR has been approved.

**Comment by YaSuenag** [comment]
Gentle ping: Could I please have review? CSR has been approved.

**Review by kevinjwalls — APPROVED** [review]
Yes, I think this looks good.
If the test ever causes problems with port availability, it could retry on new ports or use a random port, but this looks good for now.

**Comment by YaSuenag** [comment]
Thanks @kevinjwalls for the review!

> If the test ever causes problems with port availability, it could retry on new ports or use a random port, but this looks good for now.

Yeah, it is the point what I want to improve. I want to use random port for it, but I do not yet have good idea to implement it. I will try to work for it in future.

**Comment by dcubed-ojdk** [comment]
Serviceability Reviewers!
When a new SA test is added, please ask if the new test has been run with ZGC.
If it doesn't work with ZGC, then it probably needs to be added to ProblemList-zgc.txt
with an entry referring to this bug:

JDK-8220624 SA: Out of sync with ZGC's internal data structures

**Comment by plummercj** [comment]
> Serviceability Reviewers!
> When a new SA test is added, please ask if the new test has been run with ZGC.
> If it doesn't work with ZGC, then it probably needs to be added to ProblemList-zgc.txt
> with an entry referring to this bug:
> 
> JDK-8220624 SA: Out of sync with ZGC's internal data structures

Yes, I've been guilty of this many times both as a committer and as a reviewer. In fact I don't think a new SA test has ever been added and proactively problem listed for ZGC. Will try to do better.

**Comment by YaSuenag** [comment]
Sorry for my change breaks ZGC tests. I keep it in my mind.

**Pull request closed without merging**
